src/network/client.rs

```diff
@@ -130,6 +172,83 @@ impl NetworkClient {
         }
     }
 
+    async fn run_assignments_loop(&self, cancellation_token: CancellationToken) {
+        let mut timer =
+            tokio::time::interval_at(tokio::time::Instant::now(), self.assignment_update_interval);
+
+        timer.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        IntervalStream::new(timer)
+            .take_until(cancellation_token.cancelled_owned())
+            .for_each(|_| async move {
+                let latest_assignment = self
+                    .assignments
+                    .lock()
+                    .last_key_value()
+                    .map(|(assignament_id, _)| assignament_id.clone());
+                let assignment = match Assignment::try_download(
+                    self.network_state_url.clone(),
+                    latest_assignment,
+                )
+                .await
+                {
+                    Ok(Some(assignment)) => assignment,
+                    Ok(None) => {
+                        tracing::info!("Assignment has not been changed");
+                        return;
+                    }
+                    Err(err) => {
+                        tracing::error!("Unable to get assignment: {err:?}");
+                        return;
+                    }
+                };
+                let assignment_id = assignment.id.clone();
+                tracing::debug!("Got assignment {:?}", assignment_id);
+                let peers = assignment.get_all_peer_ids();
+                let mut condensed_assignment: HashMap<String, Vec<(DatasetId, Range)>> =
+                    Default::default();
+                for peer_id in peers {
+                    let mut peer_chunks: Vec<(DatasetId, Range)> = Default::default();
+                    let chunks = assignment.dataset_chunks_for_peer_id(&peer_id).unwrap();
```

Comment:
How many times is each chunk id duplicated? I believe there are multiple representations (maps) in the `Assignment` class and here we have one more copy of the entire chunk list, right? How much memory does it consume?